DOC: clean up automated tests section of workflow docs (#27696)
matplotlib/matplotlib: doc/devel/development_workflow.rst

Whenever a pull request is created or updated, various automated test tools
will run on all supported platforms and versions of Python.

* tox_ is not used in the automated testing. It is supported for testing
  locally.

  .. _tox: https://tox.readthedocs.io/

* Codecov and CodeQL are currently for information only. Their failure is not
  necessarily a blocker.

Make sure the Linting, GitHub Actions, AppVeyor, CircleCI, and Azure pipelines are
passing before merging. All checks are listed at the bottom of the GitHub page of your
pull request.

Here are some tips for finding the cause of a test failure:

- If *Linting* fails, you have a code style issue, which will be listed
  as annotations on the pull request's diff.
- If *Mypy* or *Stubtest* fails, you have inconsistency in type hints, which

>   will be listed as annotations in the diff.

Reviewer: ```suggestion
  will be listed as annotations on the diff.
```
Makes it consistent with the previous bullet

- If a GitHub Actions or AppVeyor run fails, search the log for ``FAILURES``.
  The subsequent section will contain information on the failed tests.
- If CircleCI fails, likely you have some reStructuredText style issue in
  the docs. Search the CircleCI log for ``WARNING``.
- If Azure pipelines fail with an image comparison error, you can find the
  images as *artifacts* of the Azure job:

  - Click *Details* on the check on the GitHub PR page.
  - Click *View more details on Azure Pipelines* to go to Azure.
  - On the overview page *artifacts* are listed in the section *Related*.

Skip CI checks
--------------

> If you know only a subset of CIs need to be run, this can be controlled on individual

Reviewer: ```suggestion
If you know only a subset of CI checks need to be run, this can be controlled on individual
```

commits by including the following substrings in commit messages:

> - ``[ci doc]``: restrict the CI to documentation checks. For when you only changed

Reviewer: ```suggestion
- ``[ci doc]``: restricts the CI to documentation checks. For when you have only changed
```

>   documentation. This skip is automatic if the changes are only under

Reviewer: ```suggestion
  documentation. This skip is automatic if only files changed are under
```

  ``doc/`` or ``galleries/``.
- ``[skip circle]``: skip the documentation build check. For when you didn't
  change documentation.
- Unit tests can be turned off for individual platforms with:

  - ``[skip actions]``: GitHub Actions
  - ``[skip appveyor]``: AppVeyor. Must be in the first line of the commit message.
  - ``[skip azp]``: Azure Pipelines

> - ``[skip ci]``: skip all CIs.  Use this only if you know your changes do not

Reviewer: ```suggestion
- ``[skip ci]``: skip all CI tests.  Use this only if you know your changes do not
```

>   need to be tested at all. This only skips CI triggered on ``on: push`` and

Reviewer: ```suggestion
  need to be tested at all. For the GitHub Actions CI this only skips workflows triggered on ``on: push`` and
```

  the docs. Search the CircleCI log for ``WARNING``.
- If Azure pipelines fail with an image comparison error, you can find the
  images as *artifacts* of the Azure job:

  - Click *Details* on the check on the GitHub PR page.
  - Click *View more details on Azure Pipelines* to go to Azure.
  - On the overview page *artifacts* are listed in the section *Related*.

Skip CI checks
--------------

If you know only a subset of CIs need to be run, this can be controlled on individual
commits by including the following substrings in commit messages:

- ``[ci doc]``: restrict the CI to documentation checks. For when you only changed
  documentation. This skip is automatic if the changes are only under
  ``doc/`` or ``galleries/``.
- ``[skip circle]``: skip the documentation build check. For when you didn't
  change documentation.
- Unit tests can be turned off for individual platforms with:

  - ``[skip actions]``: GitHub Actions
  - ``[skip appveyor]``: AppVeyor. Must be in the first line of the commit message.
  - ``[skip azp]``: Azure Pipelines

- ``[skip ci]``: skip all CIs.  Use this only if you know your changes do not
  need to be tested at all, which is very rare.

> Skip instructions only apply to ``on: push`` and ``on: pull_request`` events. For more

Reviewer 2: I want to be a little careful here... these terms and the link provided only apply to github actions, not appveyor/azp/circle...

Whenever a pull request is created or updated, various automated test tools
will run on all supported platforms and versions of Python.

* tox_ is not used in the automated testing. It is supported for testing
  locally.

  .. _tox: https://tox.readthedocs.io/

* Codecov and CodeQL are currently for information only. Their failure is not
  necessarily a blocker.

Make sure the Linting, GitHub Actions, AppVeyor, CircleCI, and Azure pipelines are
passing before merging. All checks are listed at the bottom of the GitHub page of your
pull request.

Here are some tips for finding the cause of a test failure:

.. list-table::
    :stub-columns: 1
    :widths: 20 15 65

    * - Linting
      - Code style
      - Errors are displayed as annotations on the pull request diff.
    * - | Mypy
        | Stubtest
      - Type hints
      - Errors are displayed as annotations on the pull request diff.
    * - CircleCI

>       - reStructuredText (rst)

Reviewer: ```suggestion
      - Documentation
```
I think this conveys what the CI is doing better than just "restructuredText"?

Reviewer 3: Can we have column titles? It may not be obvious what the columns mean.

      - Search the CircleCI log for ``WARNING``.
    * - | GitHub Actions
        | AppVeyor
        | Azure pipelines
      - Unit tests
      - | Search the log for ``FAILURES``. Subsequent section should contain information
          on failed tests.
        | On Azure, find the images as *artifacts* of the Azure job:
        |  Click *Details* on the check on the GitHub PR page.
        |  Click *View more details on Azure Pipelines* to go to Azure.
        |  On the overview page *artifacts* are listed in the section *Related*.

Skip CI checks
--------------

If you know only a subset of CI checks need to be run, you can skip any unneeded CI
checks on individual commits by including the following substrings in commit messages:

.. list-table::
    :stub-columns: 1
    :widths: 25 20 55

    * - ``[ci doc]``
      - Only run documentation checks.
      - | For when you have only changed documentation.
        | ``[ci doc]`` is applied automatically when the changes are only to files in
          ``doc/**/`` or ``galleries/**/``

>     * - ``[skip circle]``

Reviewer 3: Out of curiosity: Why is there an asymmetry between `ci doc` and `skip circle`? Would `skip doc` or `ci circle` work as well?

Reviewer 4: Looks like it would be a one line change to make `skip circle` into `skip doc`
https://github.com/matplotlib/matplotlib/blob/ff0497cc042332313afd3d2e9ba8f949c3857896/.circleci/config.yml#L20

Whenever a pull request is created or updated, various automated test tools
will run on all supported platforms and versions of Python.

* tox_ is not used in the automated testing. It is supported for testing
  locally.

  .. _tox: https://tox.readthedocs.io/

* Codecov and CodeQL are currently for information only. Their failure is not
  necessarily a blocker.

Make sure the Linting, GitHub Actions, AppVeyor, CircleCI, and Azure pipelines are
passing before merging. All checks are listed at the bottom of the GitHub page of your
pull request.

.. list-table::
    :header-rows: 1
    :stub-columns: 1
    :widths: 20 15 65

>     * - name
>       - scope
>       - tips for finding cause of failure

Reviewer 3: ```suggestion
    * - Name
      - Purpose
      - Tips for finding cause of failure
```
Would handle capitalization like section titles.

*Purpose* (=What is this for) seems more helpful than *scope*.

    * - Linting

>       - Code style

Reviewer 3: ```suggestion
      - Code style checking
```

      - Errors are displayed as annotations on the pull request diff.
    * - | Mypy
        | Stubtest

>       - Type hints

Reviewer 3: ```suggestion
      - Type checking
```
(or "Static type checking")

      - Errors are displayed as annotations on the pull request diff.
    * - CircleCI

>       - reStructuredText (rst)

Reviewer 3: ```suggestion
      - Documentation build
```

      - Search the CircleCI log for ``WARNING``.
    * - | GitHub Actions
        | AppVeyor
        | Azure pipelines
      - :ref:`tests <testing>`
      - | Search the log for ``FAILURES``. Subsequent section should contain information
          on failed tests.
        |
        | On Azure, find the images as *artifacts* of the Azure job:
        | 1. Click *Details* on the check on the GitHub PR page.
        | 2. Click *View more details on Azure Pipelines* to go to Azure.
        | 3. On the overview page *artifacts* are listed in the section *Related*.

Skip CI checks
--------------

> If you know only a subset of CI checks need to be run, you can skip any unneeded CI
> checks on individual commits by including the following substrings in commit messages:

Reviewer 3: ```suggestion
If you know only a subset of CI checks need to be run, you can skip any unneeded CI
checks on individual commits by including the following strings in commit messages:
```

"sub" does not add anything.

Also, do we want to recommend where to put these strings (first line or separate line at the end)? I'd favor the latter because these are technical control directives and not relevant when skimming through git logs. But then again, there's the exception of AppVeyor 😞. Maybe then let's not get into this.

.. list-table::
    :header-rows: 1
    :stub-columns: 1
    :widths: 25 20 55

>     * - substring
>       - effect
>       - notes

Reviewer 3: ```suggestion
    * - String
      - Effect
      - Notes
```

    * - ``[ci doc]``
      - Only run documentation checks.
      - | For when you have only changed documentation.
        | ``[ci doc]`` is applied automatically when the changes are only to files in
          ``doc/**/`` or ``galleries/**/``
    * - ``[skip doc]``
      - Skip documentation checks.
      - For when you didn't change documentation.
    * - ``[skip appveyor]``
      - Skip AppVeyor run.
      - Substring must be in first line of commit message.
    * - ``[skip azp]``
      - Skip Azure Pipelines.
      -
    * - ``[skip actions]``

>       - Skip GitHub Actions

Reviewer 3: ```suggestion
      - Skip GitHub Actions.
```
